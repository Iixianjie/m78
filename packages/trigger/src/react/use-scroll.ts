import { RefObject, useEffect, useMemo, useRef } from "react";

import { dumpFn } from "@m78/utils";
import {
  createScrollTrigger,
  type ScrollTriggerInstance,
  type ScrollTriggerOption,
  type ScrollTriggerState,
} from "../scroll/scroll.js";
import { getRefDomOrDom, useFn } from "@m78/hooks";
import { isArray } from "lodash";

/** 用于use-scroll的创建配置 */
export interface UseScrollTriggerOption
  extends Omit<ScrollTriggerOption, "target" | "handle"> {
  /** 传入要绑定的滚动元素或ref, 也可以通过useScroll返回的instance.ref绑定到dom */
  el?: HTMLElement | RefObject<any>;
  /** 滚动时触发 */
  onScroll?(meta: ScrollTriggerState): void;
}

/** use-scroll扩展后的ScrollTrigger实例 */
export interface UseScrollTriggerInstance<ElType extends HTMLElement>
  extends ScrollTriggerInstance {
  /** 可使用此项代替option.el进行绑定 */
  ref: RefObject<ElType>;
}

export function useScroll<ElType extends HTMLElement>(
  option: UseScrollTriggerOption
) {
  // 用于返回的节点获取ref
  const ref = useRef<ElType>(null);

  const instance = useMemo(() => {
    return {
      ref,
      scroll: dumpFn,
      scrollToE: dumpFn,
      scrollToElement: dumpFn,
      get: dumpFn,
      destroy: dumpFn,
    } as any as UseScrollTriggerInstance<ElType>;
  }, []);

  const handle = useFn((e: ScrollTriggerState) => {
    option.onScroll?.(e);
  });

  // 在满足条件时重载实例
  const getDeps = () => {
    const _offset = option.offset;
    const offset = isArray(_offset) ? _offset : [_offset, _offset];

    const _touchOffset = option.touchOffset;
    const touchOffset = isArray(_touchOffset)
      ? _touchOffset
      : [_touchOffset, _touchOffset];

    return [
      option.el,
      ref.current,
      (option.el as any)?.current,
      offset[0],
      offset[1],
      touchOffset[0],
      touchOffset[1],
    ];
  };

  useEffect(() => {
    const curEl = getRefDomOrDom(option.el, ref);

    if (curEl) {
      Object.assign(
        instance,
        createScrollTrigger({
          ...option,
          target: curEl,
          handle,
        })
      );
    }

    return () => {
      instance.destroy?.();
    };
  }, getDeps());

  return instance;
}
