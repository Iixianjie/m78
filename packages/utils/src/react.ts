import { AnyFunction } from "./types";

/**
 * 便捷的按键和点击事件绑定
 * @param handle - 时间处理函数
 * @param spaceTrigger - 按下空格时是否触发
 * */
export function keypressAndClick(handle: AnyFunction, spaceTrigger = true) {
  return {
    onClick: handle,
    onKeyPress: (e: any) => {
      const code = e.code;

      if (code === "Enter" || (spaceTrigger && code === "Space")) {
        handle?.();
      }
    },
  };
}
