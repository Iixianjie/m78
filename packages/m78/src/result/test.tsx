import { render } from "@testing-library/react";
import React from "react";
import { Result } from "./index.js";
import { Size } from "../common/index.js";

describe("result", function () {
  test("should render with props", function () {
    const c = render(
      <Result title="title" className="extra-cls" style={{ color: "red" }} />
    );

    expect(c.container).toMatchSnapshot();

    c.rerender(<Result title="title" desc="desc" />);

    expect(c.container).toMatchSnapshot();

    c.rerender(<Result title="title" desc="desc" extra="extra" />);

    expect(c.container).toMatchSnapshot();

    c.rerender(
      <Result title="title" desc="desc" extra="extra" actions="actions" />
    );

    expect(c.container).toMatchSnapshot();

    c.rerender(
      <Result
        title="title"
        desc="desc"
        extra="extra"
        actions="actions"
        icon="icon"
      />
    );

    expect(c.container).toMatchSnapshot();

    c.rerender(
      <Result
        title="title"
        desc="desc"
        extra="extra"
        actions="actions"
        icon="icon"
        size={Size.small}
      />
    );

    expect(c.container).toMatchSnapshot();

    c.rerender(
      <Result
        title="title"
        desc="desc"
        extra="extra"
        actions="actions"
        icon="icon"
        size={Size.large}
      />
    );

    expect(c.container).toMatchSnapshot();
  });
});
