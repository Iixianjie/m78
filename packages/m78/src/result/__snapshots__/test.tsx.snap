// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`result should render with props 1`] = `
<div>
  <div
    class="m78-result extra-cls"
    style="color: red;"
  >
    <div
      class="m78-result_title"
    >
      title
    </div>
  </div>
</div>
`;

exports[`result should render with props 2`] = `
<div>
  <div
    class="m78-result"
    style=""
  >
    <div
      class="m78-result_title"
    >
      title
    </div>
    <div
      class="m78-result_desc"
    >
      desc
    </div>
  </div>
</div>
`;

exports[`result should render with props 3`] = `
<div>
  <div
    class="m78-result"
    style=""
  >
    <div
      class="m78-result_title"
    >
      title
    </div>
    <div
      class="m78-result_desc"
    >
      desc
    </div>
    <div
      class="m78-result_extra"
    >
      extra
    </div>
  </div>
</div>
`;

exports[`result should render with props 4`] = `
<div>
  <div
    class="m78-result"
    style=""
  >
    <div
      class="m78-result_title"
    >
      title
    </div>
    <div
      class="m78-result_desc"
    >
      desc
    </div>
    <div
      class="m78-result_extra"
    >
      extra
    </div>
    <div
      class="m78-result_actions"
    >
      actions
    </div>
  </div>
</div>
`;

exports[`result should render with props 5`] = `
<div>
  <div
    class="m78-result"
    style=""
  >
    <div
      class="m78-result_icon"
    >
      icon
    </div>
    <div
      class="m78-result_title"
    >
      title
    </div>
    <div
      class="m78-result_desc"
    >
      desc
    </div>
    <div
      class="m78-result_extra"
    >
      extra
    </div>
    <div
      class="m78-result_actions"
    >
      actions
    </div>
  </div>
</div>
`;

exports[`result should render with props 6`] = `
<div>
  <div
    class="m78-result __small"
    style=""
  >
    <div
      class="m78-result_icon"
    >
      icon
    </div>
    <div
      class="m78-result_title"
    >
      title
    </div>
    <div
      class="m78-result_desc"
    >
      desc
    </div>
    <div
      class="m78-result_extra"
    >
      extra
    </div>
    <div
      class="m78-result_actions"
    >
      actions
    </div>
  </div>
</div>
`;

exports[`result should render with props 7`] = `
<div>
  <div
    class="m78-result __large"
    style=""
  >
    <div
      class="m78-result_icon"
    >
      icon
    </div>
    <div
      class="m78-result_title"
    >
      title
    </div>
    <div
      class="m78-result_desc"
    >
      desc
    </div>
    <div
      class="m78-result_extra"
    >
      extra
    </div>
    <div
      class="m78-result_actions"
    >
      actions
    </div>
  </div>
</div>
`;
